verbatimModuleSyntax + import of ambient const enum missing error

# Bug Report

<!--
  Please fill in each section completely. Thank you!
-->
[`preserveValueImports`](https://www.typescriptlang.org/tsconfig#preserveValueImports) doesn't error when you import an ambient const enum with [`isolatedModules`](https://www.typescriptlang.org/tsconfig#isolatedModules) *on*. Instead you get an error at runtime.

### :adhesive_bandage: Proposed Solution

If the named import were a type vs. an ambient const enum, you'd get a compile-time error instead.

https://github.com/microsoft/TypeScript/pull/47817 raises that error on ambient const enums, by changing its predicate to include ambient const enums.

### :mag: Search Terms

<!--
  What search terms did you use when trying to find an existing bug report?
  List them here so people in the future can find this one more easily.
-->
`preserveValueImports` ambient const enum

### :clock8: Version & Regression Information

<!-- When did you start seeing this bug occur?

"Bugs" that have existed in TS for a long time are very likely to be FAQs; refer to
  https://github.com/Microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs

If possible, please try testing the nightly version of TS to see if it's already been fixed.
For npm: `typescript@next`
This is also the 'Nightly' version in the playground: http://www.typescriptlang.org/play/?ts=Nightly

Note: The TypeScript Playground can be used to try older versions of TypeScript.

Please keep and fill in the line that best applies:
-->
4.5.0-dev.20210909 - 4.7.0-dev.20220225

### :play_or_pause_button: Playground Link

<!--
  A link to a TypeScript Playground "Share" link which shows this behavior

  The TypeScript Workbench can be used for more complex setups, try
  https://www.typescriptlang.org/dev/bug-workbench/

  As a last resort, you can link to a repo, but these will be slower for us to investigate.
-->
[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?preserveValueImports=true&isolatedModules=true#code/PTAEAEAcCcFMGdbQG6wGoEMA2BXWBJAW0gHtoAXeAKBAgEt4SsNzYATAWRLZywSrrEy5UAG9QAJRI4AdmzoyA5lzaxQAX1AAzaCUKgARACM6igHQAreAYDcVGmHBa6fGRkKwAXKBndYAfUJuXgRgE3MrYAVVAA8zNjNKKlgY0gpQAGMSGXgRWBkcfSlZeSUVNVF1IA)

### :technologist: Code

<!-- Please post the relevant code sample here as well-->
```ts repro
// @preserveValueImports
// @isolatedModules
import { RoundingMode } from "big.js";

// @filename: node_modules/big.js/index.d.ts
export const enum RoundingMode {}
```

### :slightly_frowning_face: Actual Behavior

<!-- What happened, and why it was wrong -->
Runtime error:
```
import { RoundingMode } from "big.js";
         ^^^^^^^^^^^^
SyntaxError: Named export 'RoundingMode' not found. The requested module 'big.js' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'big.js';
const { RoundingMode } = pkg;
```

### :slightly_smiling_face: Expected Behavior

<!-- What you expected to happen instead, and why -->
Compile-time error (with https://github.com/microsoft/TypeScript/pull/47817):
```
index.ts:1:10 - error TS1446: 'RoundingMode' resolves to a type-only declaration and must be imported using a type-only import when 'preserveValueImports' and 'isolatedModules' are both enabled.

1 import { RoundingMode } from "big.js";
           ~~~~~~~~~~~~
```

@andrewbranch Thanks!
- Don't we already differentiate on ambient to raise an analogous error?
  ```sh
  error TS2748: Cannot access ambient const enums when the '--isolatedModules' flag is provided.
  ```
  https://github.com/microsoft/TypeScript/blob/e4fe50cca477a3e46f9b629a6c5be0b0ed8b010f/src/compiler/checker.ts#L34251

  So already, you can't in most cases consume your own preserved const enums from a referenced project with `--isolatedModules`, and that error won't show up in the editor?
- I guess the issue with ["deconstifying" preserved const enums](https://github.com/microsoft/TypeScript/pull/46626) is if the CLI uses .d.ts files for referenced projects, it won't inline those enums ... Unless you use a ["sufficiently smart bundler":tm:](https://github.com/evanw/esbuild/issues/128#issuecomment-1002341244)?

  Discussion of dropping the const modifier from vs. adding a marker to the declaration emit of preserved const enums: https://github.com/microsoft/TypeScript/issues/46983#issue-1068952543

In other words this here proposal doesn't introduce any new issues? For the existing issue you mention, there's a solution without adding more information to declaration emit of const enums ... but it has problems with inlining referenced projects' enums? ... but maybe that can be worked around? It's a separate issue?

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -39520,7 +39520,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             error(node, Diagnostics.const_enums_can_only_be_used_in_property_or_index_access_expressions_or_the_right_hand_side_of_an_import_declaration_or_export_assignment_or_type_query);
         }
 
-        if (getIsolatedModules(compilerOptions)) {
+        // Do not check for --verbatimModuleSyntax here because imports of ambient const enums
+        // will already be checked for the same error
+        if (compilerOptions.isolatedModules) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
             const redirect = host.getRedirectReferenceForResolutionFromSourceOfProject(getSourceFileOfNode(constEnumDeclaration).resolvedPath);
@@ -46096,6 +46098,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // that inconsistency, we disallow ESM syntax in files that are unambiguously CommonJS in
                     // this mode.
                     error(node, Diagnostics.ESM_syntax_is_not_allowed_in_a_CommonJS_module_when_module_is_set_to_preserve);
+                }
+
+                if (
+                    compilerOptions.verbatimModuleSyntax &&
+                    !isTypeOnlyImportOrExportDeclaration(node) &&
+                    !(node.flags & NodeFlags.Ambient) &&
+                    targetFlags & SymbolFlags.ConstEnum
+                ) {
+                    const constEnumDeclaration = target.valueDeclaration as EnumDeclaration;
+                    const redirect = host.getRedirectReferenceForResolutionFromSourceOfProject(getSourceFileOfNode(constEnumDeclaration).resolvedPath);
+                    if (constEnumDeclaration.flags & NodeFlags.Ambient && (!redirect || !shouldPreserveConstEnums(redirect.commandLine.options))) {
+                        error(node, Diagnostics.Cannot_access_ambient_const_enums_when_0_is_enabled, isolatedModulesLikeFlagName);
+                    }
                 }
             }
 

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -39527,14 +39527,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             compilerOptions.isolatedModules
             || compilerOptions.verbatimModuleSyntax
                 && ok
-                && !(resolveName(
+                && !resolveName(
                     node,
                     getFirstIdentifier(node as EntityNameOrEntityNameExpression),
                     SymbolFlags.Alias,
                     /*nameNotFoundMessage*/ undefined,
                     /*isUse*/ false,
                     /*excludeGlobals*/ true,
-                ))
+                )
         ) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -39520,11 +39520,21 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             error(node, Diagnostics.const_enums_can_only_be_used_in_property_or_index_access_expressions_or_the_right_hand_side_of_an_import_declaration_or_export_assignment_or_type_query);
         }
 
-        // Do not check for --verbatimModuleSyntax here because imports of ambient const enums
-        // will already be checked for the same error
+        // --verbatimModuleSyntax only gets checked here when the enum usage does not
+        // resolve to an import, because imports of ambient const enums get checked
+        // separately in `checkAliasSymbol`.
         if (
-            compilerOptions.isolatedModules || compilerOptions.verbatimModuleSyntax
-                && ok && !(resolveName(node, getFirstIdentifier(node as EntityNameOrEntityNameExpression), SymbolFlags.Alias, /*nameNotFoundMessage*/ undefined, /*isUse*/ false, /*excludeGlobals*/ true))
+            compilerOptions.isolatedModules
+            || compilerOptions.verbatimModuleSyntax
+                && ok
+                && !(resolveName(
+                    node,
+                    getFirstIdentifier(node as EntityNameOrEntityNameExpression),
+                    SymbolFlags.Alias,
+                    /*nameNotFoundMessage*/ undefined,
+                    /*isUse*/ false,
+                    /*excludeGlobals*/ true,
+                ))
         ) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -39522,7 +39522,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         // Do not check for --verbatimModuleSyntax here because imports of ambient const enums
         // will already be checked for the same error
-        if (compilerOptions.isolatedModules) {
+        if (
+            compilerOptions.isolatedModules || compilerOptions.verbatimModuleSyntax
+                && ok && !(resolveName(node, getFirstIdentifier(node as EntityNameOrEntityNameExpression), SymbolFlags.Alias, /*nameNotFoundMessage*/ undefined, /*isUse*/ false, /*excludeGlobals*/ true))
+        ) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
             const redirect = host.getRedirectReferenceForResolutionFromSourceOfProject(getSourceFileOfNode(constEnumDeclaration).resolvedPath);

--- a/src/testRunner/unittests/tsc/projectReferences.ts
+++ b/src/testRunner/unittests/tsc/projectReferences.ts
@@ -115,4 +115,42 @@ describe("unittests:: tsc:: projectReferences::", () => {
             }),
         commandLineArgs: ["--p", "src/project"],
     });
+
+    verifyTsc({
+        scenario: "projectReferences",
+        subScenario: "importing const enum from referenced project with preserveConstEnums and verbatimModuleSyntax",
+        fs: () =>
+            loadProjectFromFiles({
+                "/src/preserve/index.ts": "export const enum E { A = 1 }",
+                "/src/preserve/index.d.ts": "export declare const enum E { A = 1 }",
+                "/src/preserve/tsconfig.json": jsonToReadableText({
+                    compilerOptions: {
+                        composite: true,
+                        declaration: true,
+                        preserveConstEnums: true,
+                    },
+                }),
+                "/src/no-preserve/index.ts": "export const enum E { A = 1 }",
+                "/src/no-preserve/index.d.ts": "export declare const enum F { A = 1 }",
+                "/src/no-preserve/tsconfig.json": jsonToReadableText({
+                    compilerOptions: {
+                        composite: true,
+                        declaration: true,
+                        preserveConstEnums: false,
+                    },
+                }),
+                "/src/project/index.ts": `import { E } from "../preserve";\nimport { F } from "../no-preserve";\nE.A; F.A;`,
+                "/src/project/tsconfig.json": jsonToReadableText({
+                    compilerOptions: {
+                        module: "preserve",
+                        verbatimModuleSyntax: true,
+                    },
+                    references: [
+                        { path: "../preserve" },
+                        { path: "../no-preserve" },
+                    ],
+                }),
+            }),
+        commandLineArgs: ["--p", "src/project", "--pretty", "false"],
+    });
 });

--- a/tests/baselines/reference/tsc/projectReferences/importing-const-enum-from-referenced-project-with-preserveConstEnums-and-verbatimModuleSyntax.js
+++ b/tests/baselines/reference/tsc/projectReferences/importing-const-enum-from-referenced-project-with-preserveConstEnums-and-verbatimModuleSyntax.js
@@ -1,0 +1,83 @@
+currentDirectory:: / useCaseSensitiveFileNames: false
+Input::
+//// [/lib/lib.d.ts]
+/// <reference no-default-lib="true"/>
+interface Boolean {}
+interface Function {}
+interface CallableFunction {}
+interface NewableFunction {}
+interface IArguments {}
+interface Number { toExponential: any; }
+interface Object {}
+interface RegExp {}
+interface String { charAt: any; }
+interface Array<T> { length: number; [n: number]: T; }
+interface ReadonlyArray<T> {}
+declare const console: { log(msg: any): void; };
+
+//// [/src/no-preserve/index.d.ts]
+export declare const enum F { A = 1 }
+
+//// [/src/no-preserve/index.ts]
+export const enum E { A = 1 }
+
+//// [/src/no-preserve/tsconfig.json]
+{
+  "compilerOptions": {
+    "composite": true,
+    "declaration": true,
+    "preserveConstEnums": false
+  }
+}
+
+//// [/src/preserve/index.d.ts]
+export declare const enum E { A = 1 }
+
+//// [/src/preserve/index.ts]
+export const enum E { A = 1 }
+
+//// [/src/preserve/tsconfig.json]
+{
+  "compilerOptions": {
+    "composite": true,
+    "declaration": true,
+    "preserveConstEnums": true
+  }
+}
+
+//// [/src/project/index.ts]
+import { E } from "../preserve";
+import { F } from "../no-preserve";
+E.A; F.A;
+
+//// [/src/project/tsconfig.json]
+{
+  "compilerOptions": {
+    "module": "preserve",
+    "verbatimModuleSyntax": true
+  },
+  "references": [
+    {
+      "path": "../preserve"
+    },
+    {
+      "path": "../no-preserve"
+    }
+  ]
+}
+
+
+
+Output::
+/lib/tsc --p src/project --pretty false
+src/project/index.ts(2,10): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+exitCode:: ExitStatus.DiagnosticsPresent_OutputsGenerated
+
+
+//// [/src/project/index.js]
+import { E } from "../preserve";
+import { F } from "../no-preserve";
+E.A;
+F.A;
+
+

--- a/tests/baselines/reference/verbatimModuleSyntaxAmbientConstEnum.errors.txt
+++ b/tests/baselines/reference/verbatimModuleSyntaxAmbientConstEnum.errors.txt
@@ -1,16 +1,23 @@
 /a.ts(1,10): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+/a.ts(4,1): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
 /b.ts(1,10): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
 
 
 ==== /node_modules/pkg/index.d.ts (0 errors) ====
     export declare const enum E { A, B, C }
+    declare global {
+      const enum F { A, B, C }
+    }
     
-==== /a.ts (1 errors) ====
+==== /a.ts (2 errors) ====
     import { E } from "pkg"; // Error
              ~
 !!! error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
     import type { E as _E } from "pkg"; // Ok
     console.log(E.A); // Ok
+    F.A; // Error
+    ~
+!!! error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
     
 ==== /b.ts (1 errors) ====
     export { E } from "pkg"; // Error

--- a/tests/baselines/reference/verbatimModuleSyntaxAmbientConstEnum.errors.txt
+++ b/tests/baselines/reference/verbatimModuleSyntaxAmbientConstEnum.errors.txt
@@ -1,0 +1,20 @@
+/a.ts(1,10): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+/b.ts(1,10): error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+
+
+==== /node_modules/pkg/index.d.ts (0 errors) ====
+    export declare const enum E { A, B, C }
+    
+==== /a.ts (1 errors) ====
+    import { E } from "pkg"; // Error
+             ~
+!!! error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+    import type { E as _E } from "pkg"; // Ok
+    console.log(E.A); // Ok
+    
+==== /b.ts (1 errors) ====
+    export { E } from "pkg"; // Error
+             ~
+!!! error TS2748: Cannot access ambient const enums when 'verbatimModuleSyntax' is enabled.
+    export type { E as _E } from "pkg"; // Ok
+    

--- a/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
+++ b/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
@@ -1,0 +1,12 @@
+// @verbatimModuleSyntax: true
+// @target: esnext
+// @module: preserve
+// @noEmit: true
+// @noTypesAndSymbols: true
+
+// @Filename: /node_modules/pkg/index.d.ts
+export declare const enum E { A, B, C }
+
+// @Filename: /app.ts
+import { E } from "pkg"; // Error
+console.log(E.A); // Ok

--- a/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
+++ b/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
@@ -7,6 +7,11 @@
 // @Filename: /node_modules/pkg/index.d.ts
 export declare const enum E { A, B, C }
 
-// @Filename: /app.ts
+// @Filename: /a.ts
 import { E } from "pkg"; // Error
+import type { E as _E } from "pkg"; // Ok
 console.log(E.A); // Ok
+
+// @Filename: /b.ts
+export { E } from "pkg"; // Error
+export type { E as _E } from "pkg"; // Ok

--- a/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
+++ b/tests/cases/conformance/externalModules/verbatimModuleSyntaxAmbientConstEnum.ts
@@ -6,11 +6,15 @@
 
 // @Filename: /node_modules/pkg/index.d.ts
 export declare const enum E { A, B, C }
+declare global {
+  const enum F { A, B, C }
+}
 
 // @Filename: /a.ts
 import { E } from "pkg"; // Error
 import type { E as _E } from "pkg"; // Ok
 console.log(E.A); // Ok
+F.A; // Error
 
 // @Filename: /b.ts
 export { E } from "pkg"; // Error

